File: python/src/asqav/hooks.py
"""Generic before/after hooks around ``Agent.sign(...)``.

Patterns may be glob, regex, ``"*"``, or exact action_type. Both directions
fail-open: hook exceptions are logged and signing continues.
"""

from __future__ import annotations

import fnmatch
import logging
import re
import threading
from typing import Any, Callable, Pattern, Union

logger = logging.getLogger("asqav")

PatternType = Union[str, Pattern[str]]
BeforeHook = Callable[[str, dict], Union[dict, None]]
AfterHook = Callable[[Any], None]

_lock = threading.Lock()
_before_hooks: list[tuple[PatternType, BeforeHook]] = []
_after_hooks: list[tuple[PatternType, AfterHook]] = []


def _matches(pattern: PatternType, action_type: str) -> bool:
    """Return True if ``pattern`` matches ``action_type``."""
    if isinstance(pattern, re.Pattern):
        return pattern.search(action_type) is not None
    if pattern == "*":
        return True
    if "*" in pattern or "?" in pattern or "[" in pattern:
        return fnmatch.fnmatchcase(action_type, pattern)
    return pattern == action_type


def register_before(pattern: PatternType, fn: BeforeHook) -> None:
    """Register ``fn(action_type, context)`` to run before signing matching actions.

    Return a dict to replace the context or ``None`` to keep it. Hooks run in
    registration order.
    """
    with _lock:
        _before_hooks.append((pattern, fn))


def register_after(pattern: PatternType, fn: AfterHook) -> None:
    """Register ``fn(response)`` to observe signing for matching actions."""
    with _lock:
        _after_hooks.append((pattern, fn))


def clear_hooks() -> None:
    """Remove every registered before/after hook. Intended for tests."""
    with _lock:
        _before_hooks.clear()
        _after_hooks.clear()


def _dispatch_before(action_type: str, context: dict) -> dict:
    """Invoke matching before-hooks. Returns the (possibly mutated) context."""
    with _lock:
        snapshot = list(_before_hooks)
    current = context if context is not None else {}
    for pattern, fn in snapshot:
        try:
            if not _matches(pattern, action_type):
                continue
            result = fn(action_type, current)
            if isinstance(result, dict):
                current = result
        except Exception as exc:
            logger.warning(
                "asqav before-hook %r failed (fail-open): %s", fn, exc
            )
    return current


def _dispatch_after(action_type: str, response: Any) -> None:
    """Invoke matching after-hooks. Failures are logged and swallowed."""
    with _lock:
        snapshot = list(_after_hooks)
    for pattern, fn in snapshot:
        try:
            if not _matches(pattern, action_type):
                continue
            fn(response)
        except Exception as exc:
            logger.warning(
                "asqav after-hook %r failed (fail-open): %s", fn, exc
            )

File: typescript/src/hooks.ts
/**
 * Hooks / middleware system for the Asqav TypeScript SDK.
 *
 * Mirrors the Python sibling's public surface:
 *   registerBefore(actionType, fn)  - mutate context before sign
 *   registerAfter(actionType, fn)   - observe response after sign
 *   clearHooks()                    - empty the registry
 *
 * Pattern matching:
 *   - "*"            matches every action type
 *   - "strands:*"    glob-matches any action type with the prefix "strands:"
 *   - "exact:type"   matches only that exact action type
 */

import type { SignatureResponse } from "./index.js";

export type BeforeHook = (
  actionType: string,
  context: Record<string, unknown>,
) => Record<string, unknown> | void;

export type AfterHook = (response: SignatureResponse) => void;

interface BeforeEntry {
  pattern: string;
  fn: BeforeHook;
}

interface AfterEntry {
  pattern: string;
  fn: AfterHook;
}

const beforeHooks: BeforeEntry[] = [];
const afterHooks: AfterEntry[] = [];

function patternMatches(pattern: string, actionType: string): boolean {
  if (pattern === "*") return true;
  if (pattern.endsWith(":*")) {
    const prefix = pattern.slice(0, -1); // keep the colon
    return actionType.startsWith(prefix);
  }
  if (pattern.endsWith("*")) {
    return actionType.startsWith(pattern.slice(0, -1));
  }
  return pattern === actionType;
}

export function registerBefore(actionType: string, fn: BeforeHook): void {
  beforeHooks.push({ pattern: actionType, fn });
}

export function registerAfter(actionType: string, fn: AfterHook): void {
  afterHooks.push({ pattern: actionType, fn });
}

export function clearHooks(): void {
  beforeHooks.length = 0;
  afterHooks.length = 0;
}

/** @internal */
export function _dispatchBefore(
  actionType: string,
  context: Record<string, unknown>,
): Record<string, unknown> {
  let current = context;
  for (const entry of beforeHooks) {
    if (!patternMatches(entry.pattern, actionType)) continue;
    try {
      const result = entry.fn(actionType, current);
      if (result && typeof result === "object") {
        current = result;
      }
    } catch (err) {
      // eslint-disable-next-line no-console
      console.warn(
        `[asqav] before-hook for "${entry.pattern}" threw:`,
        err instanceof Error ? err.message : err,
      );
    }
  }
  return current;
}

/** @internal */
export function _dispatchAfter(
  actionType: string,
  response: SignatureResponse,
): void {
  for (const entry of afterHooks) {
    if (!patternMatches(entry.pattern, actionType)) continue;
    try {
      entry.fn(response);
    } catch (err) {
      // eslint-disable-next-line no-console
      console.warn(
        `[asqav] after-hook for "${entry.pattern}" threw:`,
        err instanceof Error ? err.message : err,
      );
    }
  }
}
